src/minio/core/policy_builder.py

```diff
@@ -0,0 +1,189 @@
+"""
+Policy Builder module for creating and modifying MinIO policy documents.
+"""
+
+import logging
+import re
+from copy import deepcopy
+from typing import Annotated
+
+from pydantic import BaseModel, ConfigDict, Field
+
+from ...service.exceptions import PolicyOperationError
+from ..models.policy import (
+    PERMISSION_LEVEL_ACTIONS,
+    PolicyAction,
+    PolicyEffect,
+    PolicyModel,
+    PolicyPermissionLevel,
+    PolicyStatement,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class PathAccess(BaseModel):
+    """Represents access to a specific S3 path with permission level."""
+
+    model_config = ConfigDict(frozen=True)
+
+    path: Annotated[str, Field(description="S3 path for access control", min_length=1)]
+    permission_level: Annotated[
+        PolicyPermissionLevel, Field(description="Level of access permission")
+    ]
+
+
+class PolicyBuilder:
+
+    def __init__(self, policy_model: PolicyModel, bucket_name: str):
+        """
+        Initialize the PolicyBuilder with a policy model and bucket name.
+
+        Args:
+            policy_model: The base policy model to modify
+            bucket_name: The S3 bucket name for path validation
+        """
+        self.policy_model = deepcopy(policy_model)
+        self.bucket_name = bucket_name
+
+    def add_path_access(
+        self, path: str, permission_level: PolicyPermissionLevel
+    ) -> "PolicyBuilder":
+        """
+        Add access to a path with specified permission level.
+
+        Args:
+            path: The S3 path to grant access to (e.g., "s3a://bucket/path/to/data")
+            permission_level: The level of access to grant (READ, WRITE, or ADMIN)
+
+        Returns:
+            PolicyBuilder: A new builder instance with the path access added
+
+        Raises:
+            PolicyOperationError: If the path format is invalid
+        """
+        clean_path = self._normalize_path(path)
+        path_access = PathAccess(path=clean_path, permission_level=permission_level)
+
+        # Create new builder with modifications
+        new_builder = PolicyBuilder(self.policy_model, self.bucket_name)
+        new_builder._add_path_access_internal(path_access)
+        return new_builder
+
+    def build(self) -> PolicyModel:
+        """
+        Build the final policy model.
+
+        Returns:
+            PolicyModel: A deep copy of the modified policy model
+        """
+        return deepcopy(self.policy_model)
+
+    def _add_path_access_internal(self, path_access: PathAccess) -> None:
+        """Internal method to add path access to the policy."""
+        # Check for existing access to avoid duplicates
+        if self._has_path_access(path_access.path):
+            logger.warning(f"Path {path_access.path} already has access, skipping")
+            return
```

Comment:
What if the permission level is different from the current permission level? E.g. permissions are being increased or reduced for a path already in the policy?

Reply:
Ok, so there's a bug built in here that will rear its head in a later phase where you handle more permission levels. Personally, I would just fix it now since you're going to have to fix it some time, and that will make your life easier later

Reply:
The current implementation

* Can have ALLOW and DENY statements for the same resource in the same policy
* Will remove a statement with multiple paths if just one of the paths matches

That doesn't seem like behavior you want?

Reply:
Why is it ok for multiple paths to be removed?

If you're never going to use the DENY policy I"d just remove it then

Reply:
The DENY policy is defined Here: https://github.com/kbase/cdm-minio-service/pull/43/files#diff-4639ca6f7694497dc1fefd650cb10191cdea26ab2f637fc6c7c0a3a41ccbc914R15

Reply:
> So a statement will only have 1 path.

That seems fine. If that's the design, I would just throw an error if there's > 2 paths in a statement, since a design constraint has been violated

Reply:
My preference would definitely to be throw an error if my design constraints were violated, but ok

What I'd probably do is add a validator to the  PolicyStatement (or whatever) class that requires only one path

Reply:
I'm still confused why there's a DENY PolicyAction if it's never going to get used - that still means in theory there could be 2 policy statements for a path, one allowing and one denying

Reply:
> Does it matter if we have DENY PolicyAction? 

Yes, having code that is never used

* Decreases readability 
* Confuses future coders, as there's code handling cases that violate design constraints, so now they have to figure out what code is "real" and what code is extraneous
* Increases maintenance burden
* In this case allows for nonsensical state, e.g. a policy that both grants and denies access to a path

Reply:
Does it read arbitrary policies or only policies that it creates?

Reply:
👍